packages/utils/src/fastHooks.ts

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react';
+import { useRefWithInit } from './useRefWithInit';
+
+export type Instance = {
+  didInitialize: boolean;
+};
+
+type HookType = {
+  before: (instance: any) => void;
+  after: (instance: any) => void;
+};
+
+const hooks: HookType[] = [];
+
+let currentInstance: Instance | undefined = undefined;
+
+export function getInstance() {
+  return currentInstance;
+}
+
+export function setInstance(instance: Instance | undefined): void {
+  currentInstance = instance;
+}
+
+export function register(hook: HookType): void {
+  hooks.push(hook);
+}
+
+export function fastComponent<P extends object, E extends HTMLElement, R extends React.ReactNode>(
+  fn: (props: P) => R,
+): typeof fn {
+  const FastComponent = (props: P, forwardedRef: React.Ref<E>): R => {
+    const instance = useRefWithInit(createInstance).current;
+
+    let result;
+    try {
+      currentInstance = instance;
```

Comment:
WDYT about attaching the current instance to the component function instead of keeping it global to solve the concurrent rendering problem?

Reply:
What do you mean by attaching to the component function? And which concurrent rendering problem?